core/call_context.go

```diff
@@ -25,131 +25,147 @@ import (
+	if env.Depth > int(params.CallCreateDepth.Int64()) {
+		caller.ReturnGas(gas)
 
 		return nil, vm.DepthError
 	}
-	if !env.CanTransfer(caller.Address(), value) {
-		caller.ReturnGas(gas, gasPrice)
+	if !c.CanTransfer(env.StateDB, caller.Address(), value) {
+		caller.ReturnGas(gas)
 
-		return nil, ValueTransferErr("insufficient funds to transfer value. Req %v, has %v", value, env.Db().GetBalance(caller.Address()))
+		return nil, ValueTransferErr("insufficient funds to transfer value. Req %v, has %v", value, env.StateDB.GetBalance(caller.Address()))
 	}
 
-	snapshotPreTransfer := env.SnapshotDatabase()
 	var (
-		from = env.Db().GetAccount(caller.Address())
-		to   vm.Account
+		to                  vm.Account
+		snapshotPreTransfer = env.StateDB.Snapshot()
 	)
-	if !env.Db().Exist(addr) {
-		if vm.Precompiled[addr.Str()] == nil && env.ChainConfig().IsEIP158(env.BlockNumber()) && value.BitLen() == 0 {
-			caller.ReturnGas(gas, gasPrice)
+	if !env.StateDB.Exist(addr) {
+		if vm.Precompiled[addr.Str()] == nil && env.ChainConfig().IsEIP158(env.BlockNumber) && value.BitLen() == 0 {
+			caller.ReturnGas(gas)
 			return nil, nil
 		}
 
-		to = env.Db().CreateAccount(addr)
+		to = env.StateDB.CreateAccount(addr)
 	} else {
-		to = env.Db().GetAccount(addr)
+		to = env.StateDB.GetAccount(addr)
 	}
-	env.Transfer(from, to, value)
+	c.Transfer(env.StateDB, caller.Address(), to.Address(), value)
 
 	// initialise a new contract and set the code that is to be used by the
 	// EVM. The contract is a scoped environment for this execution context
 	// only.
-	contract := vm.NewContract(caller, to, value, gas, gasPrice)
-	contract.SetCallCode(&addr, env.Db().GetCodeHash(addr), env.Db().GetCode(addr))
+	contract := vm.NewContract(caller, to, value, gas)
+	contract.SetCallCode(&addr, env.StateDB.GetCodeHash(addr), env.StateDB.GetCode(addr))
 	defer contract.Finalise()
 
-	ret, err = env.Vm().Run(contract, input)
+	ret, err = env.EVM().Run(contract, input)
 	// When an error was returned by the EVM or when setting the creation code
 	// above we revert to the snapshot and consume any gas remaining. Additionally
 	// when we're in homestead this also counts for code storage gas errors.
 	if err != nil {
 		contract.UseGas(contract.Gas)
 
-		env.RevertToSnapshot(snapshotPreTransfer)
+		env.StateDB.RevertToSnapshot(snapshotPreTransfer)
 	}
 	return ret, err
 }
 
 // CallCode executes the given address' code as the given contract address
-func CallCode(env vm.Environment, caller vm.ContractRef, addr common.Address, input []byte, gas, gasPrice, value *big.Int) (ret []byte, err error) {
+func (c EVMCallContext) CallCode(env *vm.Environment, caller vm.ContractRef, addr common.Address, input []byte, gas, value *big.Int) (ret []byte, err error) {
 	// Depth check execution. Fail if we're trying to execute above the
 	// limit.
-	if env.Depth() > int(params.CallCreateDepth.Int64()) {
-		caller.ReturnGas(gas, gasPrice)
+	if env.Depth > int(params.CallCreateDepth.Int64()) {
```

Comment:
Wrong PR. This was changed in the previous PR (which this one is based on). You're right though, but unfortunately there seem to be a few tests that still test this clause so I cannot remove this yet.